internal/controllers/clusterextension_controller.go

```diff
@@ -17,78 +17,121 @@ limitations under the License.
+	"github.com/operator-framework/rukpak/pkg/util"
 
 	ocv1alpha1 "github.com/operator-framework/operator-controller/api/v1alpha1"
 	"github.com/operator-framework/operator-controller/internal/catalogmetadata"
 	catalogfilter "github.com/operator-framework/operator-controller/internal/catalogmetadata/filter"
 	catalogsort "github.com/operator-framework/operator-controller/internal/catalogmetadata/sort"
+	"github.com/operator-framework/operator-controller/internal/conditionsets"
+	"github.com/operator-framework/operator-controller/internal/handler"
+	"github.com/operator-framework/operator-controller/internal/labels"
 )
 
 // ClusterExtensionReconciler reconciles a ClusterExtension object
 type ClusterExtensionReconciler struct {
 	client.Client
-	BundleProvider BundleProvider
+	ReleaseNamespace   string
+	BundleProvider     BundleProvider
+	Unpacker           rukpaksource.Unpacker
+	ActionClientGetter helmclient.ActionClientGetter
+	Storage            storage.Storage
+	Handler            handler.Handler
+	Scheme             *runtime.Scheme
+	dynamicWatchMutex  sync.RWMutex
+	dynamicWatchGVKs   map[schema.GroupVersionKind]struct{}
+	controller         crcontroller.Controller
+	cache              cache.Cache
 }
 
 //+kubebuilder:rbac:groups=olm.operatorframework.io,resources=clusterextensions,verbs=get;list;watch
 //+kubebuilder:rbac:groups=olm.operatorframework.io,resources=clusterextensions/status,verbs=update;patch
 //+kubebuilder:rbac:groups=olm.operatorframework.io,resources=clusterextensions/finalizers,verbs=update
-
-//+kubebuilder:rbac:groups=core.rukpak.io,resources=bundledeployments,verbs=get;list;watch;create;update;patch
+//+kubebuilder:rbac:groups=core,resources=pods,verbs=list;watch;create;delete
+//+kubebuilder:rbac:groups=core,resources=configmaps,verbs=list;watch
+//+kubebuilder:rbac:groups=core,resources=pods/log,verbs=get
+//+kubebuilder:rbac:groups=*,resources=*,verbs=*
 
 //+kubebuilder:rbac:groups=catalogd.operatorframework.io,resources=catalogs,verbs=list;watch
 //+kubebuilder:rbac:groups=catalogd.operatorframework.io,resources=catalogmetadata,verbs=list;watch
 
+// The operator controller needs to watch all the bundle objects and reconcile accordingly. Though not ideal, but these permissions are required.
+// This has been taken from rukpak, and an issue was created before to discuss it: https://github.com/operator-framework/rukpak/issues/800.
 func (r *ClusterExtensionReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
 	l := log.FromContext(ctx).WithName("operator-controller")
 	l.V(1).Info("starting")
 	defer l.V(1).Info("ending")
 
 	var existingExt = &ocv1alpha1.ClusterExtension{}
-	if err := r.Get(ctx, req.NamespacedName, existingExt); err != nil {
-		return ctrl.Result{}, client.IgnoreNotFound(err)
+	if err := r.Client.Get(ctx, req.NamespacedName, existingExt); err != nil {
+		return ctrl.Result{}, utilerrors.NewAggregate([]error{client.IgnoreNotFound(err), nil})
```

Comment:
+1. Not sure why this was changed from simply returning `client.IgnoreNoteFound(err)`

Reply:
See #878

Reply:
#878 is on hold until this PR merges, and #878 will deal with _all_ the error returns. So, any discussions regarding error returns can be done on that PR when it's updated. i.e. we're deferring error handling fixups (to make it easier to review with less churn).

Reply:
> #878 is on hold until this PR merges, and #878 will deal with _all_ the error returns. So, any discussions regarding error returns can be done on that PR when it's updated. i.e. we're deferring error handling fixups (to make it easier to review with less churn). #878 will be relevant when complete.

But... There is a bug in this line: when object is 404 we just need to return `nil` in the error. With this change - we will always be returning a non-nil error. Unless I'm missing something, this will lead to a constant re-queuing of 404 objects.

Also for some reason there is an extra `nil` in the slice of errors:

```go
return ctrl.Result{}, utilerrors.NewAggregate([]error{client.IgnoreNotFound(err), nil})
//                                                                                ^
```

Reply:
>There is a bug in this line: when object is 404 we just need to return nil in the error. With this change - we will always be returning a non-nil error. Unless I'm missing something, this will lead to a constant re-queuing of 404 objects.

My understanding is that this will not always return a non-nil error. If `client.IgnoreNotFound(err)` returns `nil` then all the errors in the list passed to `utilerrors.NewAggregate` are `nil` and `utilerrors.NewAggregate` will return a `nil` error.

Ref: https://github.com/kubernetes/apimachinery/blob/37988e577e160bf7749437141fee1d6b8ca15dd6/pkg/util/errors/errors.go#L46-L61

Reply:
I agree with @tmshort that we should move discussion to the new PR tailored specifically to address all error issues as a follow-up to this PR.

Reply:
You are right. I did not look into what `NewAggregate` does and assumed it just packs a slice of errors into own type.

Reply:
Resolving this conversation for now.